Explain the bridge between SF and legacy hooks

matks opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop developer documentation! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines on how to contribute:
https://devdocs.prestashop.com/1.7/contribute/documentation/how/
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | Explain the bridge between SF and legacy hooks

matks commented:
PR ready for review

jolelievre commented on src/content/1.7/development/architecture/legacy/hooks.md:
````diff
@@ -0,0 +1,89 @@
+$id = Hook::exec('actionModifyZoning', ['address_id' => $addressID]);
+```
+
+The modules which have previously subscribed to this hook will be called and will be able act on this hook. Depending on the hook nature, they can alter the data being passed, trigger an event of even return a result (a data structure or raw HTML) to be used within PrestaShop.
````
```suggestion
The modules which have previously subscribed to this hook will be called and will be able to act on this hook. Depending on the hook nature, they can alter the data being passed, trigger an event or even return a result (a data structure or raw HTML) to be used within PrestaShop.
```

jolelievre commented on src/content/1.7/development/architecture/legacy/hooks.md:
```diff
@@ -0,0 +1,89 @@
+
+First, on setup, class LegacyHookSubscriber will retrieve all legacy hooks using `hooks = Hook::getHooks();`.
+
+Then it will create as many event subscriptions as there are hooks, following a naming convention.
```
```suggestion
Then it will create as many event subscriptions as there are hooks, following a naming convention based on database IDs.
```

jolelievre commented on src/content/1.7/development/architecture/legacy/hooks.md:
```diff
@@ -0,0 +1,89 @@
+
+Then it will create as many event subscriptions as there are hooks, following a naming convention.
+
+For example if module 267 has subscribed to hook 82, LegacyHookSubscriber will create a subscription `call_82_267`.
```
```suggestion
For example if module with `id_module` 267 has subscribed to hook with `id_hook` 82, `LegacyHookSubscriber` will create a subscription `call_82_267`.
```

jolelievre commented on src/content/1.7/development/architecture/legacy/hooks.md:
```diff
@@ -0,0 +1,89 @@
+
+The HookDispatcher will act as standard dispatcher and call the eligible event listeners and subscribers, including LegacyHookSubscriber.
+
+Using a dedicated magic method `__call()`, the LegacyHookSubscriber will track which hook has been called and trigger the related `Hook::exec()` call.
```
```suggestion
Using a dedicated magic method `__call()`, the `LegacyHookSubscriber` will parse the called method (ex: `call_82_267`) to retrieve the appropriate hook and module (using their respective IDs) and trigger the related `Hook::exec()` call with the appropriate parameters.
```

matks commented:
> I think this article should be split in two:
> 
> The first part should describe Hooks as a [PrestaShop component](https://devdocs.prestashop.com/1.7/development/components/).
> 
> The second one could be a subsection of the other page, explaining how the bridge between events and hooks works. A link to this article should be added [here](https://devdocs.prestashop.com/1.7/development/architecture/migration-guide/hooks/#dispatching-hooks-using-the-hook-dispatcher) as welll

Done in https://github.com/PrestaShop/docs/pull/661/commits/0691892c8df4a4abef4f1b5edb73030286fa656d

- new "Hooks component" section
- new "Symfony bridge for hooks" page in this section
- links to Hook concept now target this section
- I also moved "Register a new hook" how-to into this section and added an alias to preserve URL

eternoendless commented on src/content/1.7/development/components/hooks/symfony-bridge.md:
```diff
@@ -0,0 +1,47 @@
+
+Then it will create as many event subscriptions as there are hooks, following a naming convention based on database IDs.
+
+For example if module with `id_module` 267 has subscribed to hook with `id_hook` 82, `LegacyHookSubscriber` will create a subscription `call_82_267`.
```
```suggestion
For example, if module with `id_module` 267 has subscribed to hook with `id_hook` 82, `LegacyHookSubscriber` will create a subscription `call_82_267`.
```

eternoendless commented on src/content/1.7/development/components/hooks/symfony-bridge.md:
```diff
@@ -0,0 +1,47 @@
+
+## To sum up
+
+When a hook is dispatched inside Symfony-powered pages of the Back-office, the HookDispatcher dispatches the hook similarly to a regular Symfony event, and this event is being listened to by LegacyHookSubscriber that will then trigger the related `Hook::exec()` call.
```
```suggestion
When a hook is dispatched inside Symfony-powered controllers, `HookDispatcher` dispatches the hook similarly to a regular Symfony event; this event is being listened to by `LegacyHookSubscriber` who will then trigger the related `Hook::exec()` call.
```

matks commented on src/content/1.7/development/components/hooks/_index.md:
```diff
@@ -0,0 +1,70 @@
+
+The main way to customize PrestaShop is using Modules. Modules allow to customize PrestaShop in many ways.
+
+The main path for Module integration is Hooks, which are placed throughout the system. Modules can subscribe to hooks in order to provide or alter features.
```
For the record 😛  I stole these words from yourself in https://build.prestashop.com/news/prestashop-in-2019-and-beyond-part-1-current-architecture/

eternoendless replied:
You forgot that I had written this two paragraphs before 😉

> The Modules system provides a plug-in approach to added functionality. As explained before, it relies mainly on specified extension points called “Hooks”

matks commented on src/content/1.7/development/components/hooks/register-new-hook.md:
```diff
@@ -21,7 +21,7 @@ final class FormHandler extends AbstractFormHandler
     /* [...] */
     public function getForm()
-    {
+    {s
```
Ouuups

matks commented on src/content/1.7/modules/concepts/hooks/_index.md:
```diff
@@ -125,3 +125,5 @@ $this->registerHook('displayAtSpecificPlace');
 
 If the hook "displayAtSpecificPlace" doesn't exist, PrestaShop will create it for you but be carefull : this will also plug the current module to the hook.
+
+[hooks-component]: {{< ref "/1.7/development/components/hooks/_index.md" >}}
```
```suggestion
[hooks-component]: {{< ref "/1.7/development/components/hook/_index.md" >}}
```

matks commented on src/content/1.7/development/architecture/migration-guide/hooks.md:
````diff
@@ -76,3 +76,5 @@ Some hooks are directly rendered in templates, because PrestaShop developers wan
 ) }}
 ```
+
+[hooks-component]: {{< ref "/1.7/development/components/hooks/_index.md" >}}
````
```suggestion
[hooks-component]: {{< ref "/1.7/development/components/hook/_index.md" >}}
```

matks commented:
@eternoendless All feedbacks have been applied. Unfortunately github is behaving in weird ways right now, so I had to apply feedbacks in small batches as my "apply suggestions". That created multiple git commits so I squashed all that.

eternoendless commented on src/content/1.7/development/components/hook/_index.md:
````diff
@@ -0,0 +1,72 @@
+$id = Hook::exec('actionModifyZoning', ['address_id' => $addressID]);
+```
+
+## Modules callback
````
```suggestion
## Module callback
```

eternoendless commented on src/content/1.7/development/components/hook/register-new-hook.md:
```diff
@@ -1,6 +1,8 @@
 ---
 title: Register a new hook
 weight: 60
```
```suggestion
weight: 10
```

eternoendless commented on src/content/1.7/development/components/hook/symfony-bridge.md:
```diff
@@ -0,0 +1,44 @@
+---
+title: Symfony bridge for hooks
```
```suggestion
title: Symfony bridge for hooks
weight: 20
```

eternoendless commented on src/content/1.7/development/components/hook/symfony-bridge.md:
```diff
@@ -0,0 +1,44 @@
+
+In Back-Office, migrated pages are powered by Symfony. The powerful [Symfony Event Dispatcher][sf-event-dispatcher] has been chosen to replace, in the long term, hooks. However, in order to preserve backward compatibility, both systems coexist in 1.7.
+
+It is important that modules which subscribe to a hook are correctly notified no matter where the hook is dispatched form, be it in legacy code or modern code.
```
```suggestion
It is important that modules which subscribe to a hook are correctly notified no matter where the hook is dispatched from, be it in legacy code or modern code.
```

eternoendless commented:
Closed/reopened to trigger Travis

eternoendless commented:
Thanks @matks

matks commented:
🤩
